streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsKafkaClient.java

```diff
@@ -0,0 +1,216 @@
+        for (String key : topicProperties.stringPropertyNames()) {
+            topicConfig.put(key, topicProperties.getProperty(key));
+        }
+        final CreateTopicsRequest.TopicDetails topicDetails = new CreateTopicsRequest.TopicDetails(numPartitions, (short) replicationFactor, topicConfig);
+        final Map<String, CreateTopicsRequest.TopicDetails> topics = new HashMap<>();
+        topics.put(internalTopicConfig.name(), topicDetails);
+
+        final CreateTopicsRequest createTopicsRequest = new CreateTopicsRequest(topics, streamsConfig.getInt(StreamsConfig.REQUEST_TIMEOUT_MS_CONFIG));
+
+        final ClientResponse clientResponse = sendRequest(createTopicsRequest.toStruct(), ApiKeys.CREATE_TOPICS);
+        final CreateTopicsResponse createTopicsResponse = new CreateTopicsResponse(clientResponse.responseBody());
+        if (createTopicsResponse.errors().get(internalTopicConfig.name()).code() > 0) {
+            throw new StreamsException("Could not create topic: " + internalTopicConfig.name() + ". " + createTopicsResponse.errors().get(internalTopicConfig.name()).name());
+        }
+    }
+
+
+    /**
+     * Send a request to kafka broker of this client. Polls the request for a given number of iterations to receive the response.
+     *
+     * @param request
+     * @param apiKeys
+     */
+    private ClientResponse sendRequest(final Struct request, final ApiKeys apiKeys) {
+
+        final Node brokerNode = kafkaClient.leastLoadedNode(new SystemTime().milliseconds());
+        final String brokerId = Integer.toString(brokerNode.id());
+
+        final SystemTime systemTime = new SystemTime();
+
+        final RequestSend send = new RequestSend(brokerId,
+                kafkaClient.nextRequestHeader(apiKeys),
+                request);
+
+        final ClientRequest clientRequest = new ClientRequest(systemTime.milliseconds(), true, send, null);
+
+
+        final long readyTimeout = systemTime.milliseconds() + MAX_WAIT_TIME_MS;
+        while (!kafkaClient.ready(brokerNode, systemTime.milliseconds()) && systemTime.milliseconds() < readyTimeout) {
+            kafkaClient.poll(streamsConfig.getLong(StreamsConfig.POLL_MS_CONFIG), systemTime.milliseconds());
+        }
+        if (!kafkaClient.ready(brokerNode, systemTime.milliseconds())) {
+            throw new StreamsException("Timed out waiting for node=" + brokerNode + " to become available");
+        }
+        kafkaClient.send(clientRequest, systemTime.milliseconds());
+
+        final long responseTimeout = systemTime.milliseconds() + MAX_WAIT_TIME_MS;
+        // Poll for the response.
+        while (systemTime.milliseconds() < responseTimeout) {
+            List<ClientResponse> responseList = kafkaClient.poll(streamsConfig.getLong(StreamsConfig.POLL_MS_CONFIG), systemTime.milliseconds());
+            for (ClientResponse clientResponse: responseList) {
+                if (clientResponse.request().equals(clientRequest)) {
```

Comment:
Is it a good behavior that we are simply dropping all other responses on the floor while waiting for the corresponding response? I think today we will not encounter such issues since we always send one request, and block on its response and then send another one. But this is less efficient since with N topics to create we have to go N round trips now, suppose moving forward we will do that in a more batched manner where multiple in-flight requests exist, then I this will be an issue.

So instead of checking if the response, could we use the `RequestCompletionHandler` interface with sth. similar to consumer where we poll until the handler set the `Future` indicating it is received and processed?